criu/mount.c

```diff
@@ -929,6 +929,28 @@ static void search_bindmounts(void)
 		__search_bindmounts(mi);
 }
 
+struct mount_info *mnt_bind_pick(struct mount_info *mi, bool (*pick)(struct mount_info *mi, struct mount_info *bind))
+{
+	struct mount_info *bind;
+
+	BUG_ON(!mi);
+
+	if (pick(mi, mi))
+		return mi;
+
+	/*
+	 * Shouldn't use mnt_bind list before it was setup in
+	 * __search_bindmounts
+	 */
+	BUG_ON(list_empty(&mi->mnt_bind) && !mi->mnt_no_bind);
```

Comment:
The mi->mnt_no_bind==true is an indicator that empty mi->mnt_bind list reflects that mount has no bindmounts other than itself. If both mnt_no_bind is false and list is empty than bindmount search haven’t been done yet and one can’t yet use the list

Reply:
That's a lot of information stuffed into that little line of code. :) Maybe a tiny static inline helper with a comment explaining that would help. But just a thought.

Reply:
I've just simplified the code a bit. Now we have mnt_bind_is_populated instead of mnt_no_bind.

The mnt_no_bind was set only for mounts with empty mnt_bind list, it was a bit more effective in terms of memory writes, but mnt_bind_is_populated is more simple, mnt_bind_is_populated now is set for all mounts just after we've collected bindmounts for them, so proper bug_on is simple as `BUG_ON(!m->mnt_bind_is_populated);`